### PR TITLE
Sane numpy

### PR DESCRIPTION
**Pull request by ocefpaf**
@kain88-de  hopefully this one does the "right" thing.

Alternative to #56

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,57 +1,57 @@
...
   sha256: fc4de6feaa6682801abac19f8e4f3ed0ea56b39e7e5a89ab66c7e0a76f661445
   patches:
     # TODO: Remove this patch when 0.19 is released
     - skip-tests-failing-on-0.18.2.patch
 
 build:
-  number: 200
+  number: 201
   # We lack openblas on Windows, and therefore can't build numpy there either currently.
-  skip: true  # [win or np!=113]
```
Comment (ocefpaf) [inline recipe/meta.yaml]: @kain88-de this is the problem you had there. Vanilla `conda-smithy` was doing the right thing and obeying the skip, hence the deletion of the CI config in #56. 

My hacked version is more verbose and forces what the user wants, causing an error when trying to build. I prefer that b/c I use it to debug issues.

**Comment by ocefpaf** [comment]
If this passes in one go we can remove the branches.

File: recipe/meta.yaml (hunk 1 above)
Comment (kain88-de) [inline recipe/meta.yaml]: The line seems to be there because this recipy manages numpy versions with branches.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,57 +1,57 @@
...
   build:
-    - cython
     - python
+    - cython
     - setuptools
     - blas 1.1 {{ variant }}
     - openblas 0.2.19|0.2.19.*
-    - numpy x.x
+    - numpy 1.8.*  # [not (win and py36)]
+    - numpy 1.11.*  # [win and py36]
```
Comment (jakirkham) [inline recipe/meta.yaml]: Why the difference between Windows and non-Windows?

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,57 +1,57 @@
 {% set version = "0.18.2" %}
-
 {% set variant = "openblas" %}
 
 package:
   name: scikit-learn
   version: {{ version }}
 
 source:
-  fn: scikit-learn-{{ version }}.tar.gz
```
Comment (jakirkham) [inline recipe/meta.yaml]: This shouldn't be deleted right?

File: recipe/meta.yaml (hunk 1 above)
Comment (ocefpaf) [inline recipe/meta.yaml]: Yep. But if this works we no longer need the multiple branches. I need to check the minimum version of `numpy` required by `scikit-learn` though. Maybe it is not np18 :-/

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,57 +1,57 @@
...
     - scipy
     - nose
-
   run:
     - python
     - blas 1.1 {{ variant }}
     - openblas 0.2.19|0.2.19.*
-    - numpy x.x
+    - numpy >=1.8  # [not (win and py36)]
+    - numpy >=1.11  # [win and py36]
```
Comment (jakirkham) [inline recipe/meta.yaml]: Same question as above.

File: recipe/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipe/meta.yaml]: Actually the Windows line shouldn't even be here if we are still skipping building on Windows.

File: recipe/meta.yaml (hunk 3 above)
Comment (ocefpaf) [inline recipe/meta.yaml]: Why not? Embrace the future.

File: recipe/meta.yaml (hunk 4 above)
Comment (ocefpaf) [inline recipe/meta.yaml]: See https://conda-forge.org/docs/meta.html#building-against-numpy

File: recipe/meta.yaml (hunk 2 above)
Comment (ocefpaf) [inline recipe/meta.yaml]: > Actually the Windows line shouldn't even be here if we are still skipping building on Windows.

You are the first one to ensure first as future proof. I don't see why not leave them there. Numpy with openblas for Windows will be coming to conda-forge sooner than you think.

**Comment by kain88-de** [comment]
The error you are hitting compiling is legitimate and has been removed recently https://github.com/scikit-learn/scikit-learn/pull/9129

I wonder why it never showed up before. Maybe using an old cython version could "fix" this.

**Comment by ocefpaf** [comment]
> The error you are hitting compiling is legitimate and has been removed recently https://github.com/scikit-learn/scikit-learn/pull/9129

Thanks for the info! I found out another one too. See the two patches I added here.

> I wonder why it never showed up before. Maybe using an old cython version could "fix" this.

I tested and an older `cython` also works but, b/c those are fixed upstream, I decided to just backport the patches.

**Comment by kain88-de** [comment]
The circle I build timed out. Maybe running the full test suite is a bit much.

**Comment by ocefpaf** [comment]
> The circle I build timed out. Maybe running the full test suite is a bit much.

It is. But I am testing another new thing we added to `conda-smithy` to handle that. Let's see how `CircleCI 2.0` matrix fairs here. (We still need to fix the ugly fast finish stuff though.)

**Comment by amueller** [comment]
is this about using a new conda-smithy version? I'm not really following what's going on.

**Comment by ocefpaf** [comment]
> is this about using a new conda-smithy version?

Not really but we ended up using the latest `conda-smithy` too in order to get a "matrix" for `CircleCI` so we can remove the branches used to build `scikit-learn`. Now we can build everything in one go.

> I'm not really following what's going on.

The `numpy` part is, in essence, the same approach used to builds wheels. We build with the oldest `numpy` possible so we can run with any  newer version that is backwards compatible.

Note that this approach reduces the build matrix dramatically (from 9 to 3 items), and ensure that we don't have to rebuild `scikit-learn` whenever a new version of `numpy` comes out. (Only if there is a change on the `numpy` ABI that breaks backward compatibility, then we do have to build it again.)

**Comment by amueller** [comment]
@ocefpaf got it, makes a lot of sense. thanks for the summary!

**Comment by ocefpaf** [comment]
No problem. Feel free to oping me if you have more questions.

**Pull request merged**